src/core/net/srp_client.cpp

```diff
@@ -0,0 +1,1551 @@
+    // `ItemState` to define transitions between states in `Pause()`,
+    // `Stop()`, `SendUpdate`, and `ProcessResponse()`, or to convert
+    // an `ItemState` to string. Here, we assert that the enumeration
+    // values are correct.
+
+    static_assert(kToAdd == 0, "kToAdd value is not correct");
+    static_assert(kAdding == 1, "kAdding value is not correct");
+    static_assert(kToRefresh == 2, "kToRefresh value is not correct");
+    static_assert(kRefreshing == 3, "kRefreshing value is not correct");
+    static_assert(kToRemove == 4, "kToRemove value is not correct");
+    static_assert(kRemoving == 5, "kRemoving value is not correct");
+    static_assert(kRegistered == 6, "kRegistered value is not correct");
+    static_assert(kRemoved == 7, "kRemoved value is not correct");
+}
+
+otError Client::Start(const Ip6::SockAddr &aServerSockAddr, Callback aCallback, void *aContext)
+{
+    otError error = OT_ERROR_NONE;
+
+    VerifyOrExit(GetState() == kStateStopped, error = OT_ERROR_ALREADY);
+
+    SuccessOrExit(error = mSocket.Open(Client::HandleUdpReceive, this));
+    SuccessOrExit(error = mSocket.Connect(aServerSockAddr));
+
+    otLogInfoSrp("[client] Starting, server [%s]:%d", aServerSockAddr.GetAddress().ToString().AsCString(),
+                 aServerSockAddr.mPort);
+
+    mCallback        = aCallback;
+    mCallbackContext = aContext;
+    Resume();
+
+exit:
+    return error;
+}
+
+void Client::Stop(void)
+{
+    // Change the state of host info and services so that they are
+    // added/removed again once the client is started back. In the
+    // case of `kAdding`, we intentionally move to `kToRefresh`
+    // instead of `kToAdd` since the server may receive our add
+    // request and the item may be registered on the server. This
+    // ensures that if we are later asked to remove the item, we do
+    // notify server.
+
+    static const ItemState kNewStateOnStop[]{
+        /* (0) kToAdd      -> */ kToAdd,
+        /* (1) kAdding     -> */ kToRefresh,
+        /* (2) kToRefresh  -> */ kToRefresh,
+        /* (3) kRefreshing -> */ kToRefresh,
+        /* (4) kToRemove   -> */ kToRemove,
+        /* (5) kRemoving   -> */ kToRemove,
+        /* (6) kRegistered -> */ kToRefresh,
+        /* (7) kRemoved    -> */ kRemoved,
+    };
+
+    VerifyOrExit(GetState() != kStateStopped);
+
+    // State changes:
+    //   kAdding     -> kToRefresh
+    //   kRefreshing -> kToRefresh
+    //   kRemoving   -> kToRemove
+    //   kRegistered -> kToRefresh
+
+    mHostInfo.SetState(kNewStateOnStop[mHostInfo.GetState()]);
+
+    for (Service *service = mServices.GetHead(); service != nullptr; service = service->GetNext())
+    {
+        service->SetState(kNewStateOnStop[service->GetState()]);
+    }
+
+    IgnoreError(mSocket.Close());
+    mShouldRemoveKeyLease = false;
+    mTxFailureRetryCount  = 0;
+    ResetRetryWaitInterval();
+    SetState(kStateStopped);
+
+exit:
+    return;
+}
+
+void Client::Resume(void)
```

Comment:
Should we also call `ResetRetryWaitInterval()` here?

Reply:
`Pause` and `Resume` are used when device detaches  and then attaches  again.

I am not sure about this. I see value in both situations (resetting the wait interval or not resetting it):
- On one hand, I think we should resume as we were before (so if we were in middle of retries and back-to-back failures, then continue with whatever retry interval we were using befor)
  -  When we `Resume` we will try sending an update immediately (if there are pending/unregistered changes) but if it fails we go to same retry interval schedule as before.
- On other hand, if we detach and then attach again (maybe to a new parent), we may now have a path to server and should try again harder, so resetting the retry interval make sense in this case

For now I left it as is (not reseting the retry wait interval).

@jwhui  Thoughts?

Reply:
I had the same tradeoff in mind. I'm fine with returning to the existing retry interval for now. However, one corner case to consider is that we should avoid quickly growing the retry interval simply because we detach/reattach many times in succession. In other words, I think we should only grow the retry interval once the current retry interval has expired. Thoughts?

Reply:
Yes. This should work correctly as is, since we grow the retry interval after time out from `HandleTimer`.
If we are paused while waiting for a response, when we resume, we prepare and send an update message and again re-use the same wait interval.

Reply:
Consider the following:
1. Send update message.
2. Response timeout.
3. Grow retry wait interval.
4. Detach
5. Reattach
6. Repeat 1.

Will we grow the retry wait interval more quickly than expected in the above scenario?

Reply:
I think this should be fine. 
 
Note that the timeout interval we use after a tx is actually the current retry wait interval (wait time till the next retry). 

So at end of step 2  "Response timeout" we know that we have have waited for the current retry wait interval already (and received no response), we then grow the retry interval and go back to `kStateToUpdate` state to then prepare and send a new update message (now waiting for the longer/new wait interval). If we are detached after this, we correctly resume with the same longer/new wait interval.

Reply:
Ah, yes. Thanks for the clarification. Thanks.